### ISSUE TITLE
Improve syntax error warnings in various situations

### ISSUE THREAD

**Issue by justinbmeyer** [stealjs/steal#1378]
## Main module error

```js
// breaking.js
thing broken;
```

```html
<script src="../node_modules/steal/steal.js" main="~/breaking"></script>
```

Result:

![can-connect_tests-6](https://user-images.githubusercontent.com/78602/38567468-ee4e9542-3cab-11e8-9427-7e1f39702d28.png)

## Importing a module with a syntax error

```js
// importer.js
require("./breaking");
```

```html
<script src="../node_modules/steal/steal.js" main="~/importer"></script>
```

Result:

![can-connect_tests-7](https://user-images.githubusercontent.com/78602/38567589-3b0fcefa-3cac-11e8-8203-791e6decb2d8.png)

But the file is link-able:

![can-connect_tests-8](https://user-images.githubusercontent.com/78602/38567620-5279bc72-3cac-11e8-8b90-f75efc21e099.png)

**Comment by matthewp** [comment]
I'll add this to my bucket of things to get to. This should show the file where the error occurs...

**Comment by matthewp** [comment]
The problem is that when a syntax error is thrown in the dependant module that file is not part of the stack trace. So when steal goes to add the codeframe to the error, it thinks that `importer.js` is the source of the error and adds a codeframe for that.

We need to inject `breaking.js` into the stack trace, which I think should prevent the wrong codeframe from being added.

I'm trying to figure out how Chrome manages to show the true source of the error when not using `try/catch`. That information is not part of the stack trace. There is Chrome-specific [APIs](https://github.com/v8/v8/wiki/Stack-Trace-API) for errors, so maybe they are using one of those. Investigating.

**Comment by justinbmeyer** [comment]
@matthewp since it's a syntax error .. I'm betting we could do the following ... 

Let steal run the normal way ... it will create a syntax error, but since we are catching it, we can't get the actual line number.

But since we know it's a syntax error, run the code again without a try / catch, letting the syntax error show up to the user.  

Steal will never see it, but it doesn't have to.

**Comment by matthewp** [comment]
That's an interesting idea, I can try that tomorrow.

**Comment by matthewp** [comment]
0ea3618 prevents the wrong error from being logged, reopening to try out the re-eval trick.

**Comment by matthewp** [comment]
I think what I might do here is create an API without steal so that errors can control how they are logged. When we log the error we can check for a `error.logError()` function, and if that exists use it.

```js
if(typeof error.logError = "function") {
  error.logError();
} else {
  console.error(error);
}
```

Example implementation:

```js
if(error instanceof SyntaxError) {
  error.logError = function(){
    // Using setTimeout to prevent being swallowed within a promise.
    setTimeout(function(){
      new Function(source);
    });
  };
}
```

**Comment by matthewp** [comment]
I've implemented this idea in [this branch](https://github.com/stealjs/steal/tree/syntax-throw). It looks like below:

![syntax-err](https://user-images.githubusercontent.com/361671/38686228-e03e962c-3e41-11e8-8cdd-bdd5d128f7e8.png)

Some things to think about:

1. This does provide a link to the error in the code, whereas in the other implementation we couldn't get the line number.
2. Because we rely on letting Chrome catch the error, we don't have the rest of the call stack in the stack trace. We only have steal (which we don't want in there).
3. There's an inconsistency with other errors, where the source of the error is the first item in the stack trace. In this case the stack trace is not used, but instead you look at the link on the right.
4. This method only works in Chrome; in Firefox this is worth than just providing a stack trace without the line number. So if we go this way we need to special-case it for Chrome.

**Comment by justinbmeyer** [comment]
If you clone / npm install this branch: https://github.com/canjs/can-query-logic/tree/missing-comma

And go to:  http://localhost:8080/can-query-logic/test.html

You'll see something like:

![_can-query-logic_and_canjs](https://user-images.githubusercontent.com/78602/39000709-4a3d18e6-43ba-11e8-96ad-a7c59a853311.png)

This is because there's a missing comma in `src/types/comparisons-test.js` right before `NotIn_NotIn`'s tests.

**Comment by matthewp** [comment]
Looking into this now. There's a problem with the way catching works that makes this tricky. But basically what happens is we create new Errors after catching the root one, but don't use `error.constructor` so the special logic for SyntaxError types doesn't get used.

That part is easy to fix. But even with that fix there's still the problem of not knowing where the syntax error occurs. In long files (like this one) with a small error like a missing comma it can be *really* hard to find.

With that being said, I think we do have to implement [this](https://github.com/stealjs/steal/issues/1378#issuecomment-380841218) even though the downsides listed in that comment still hold true.

**Comment by matthewp** [comment]
This is fixed in steal 1.11.6. Now we show the line number to the right of the error like below:

<img width="709" alt="screen shot 2018-05-01 at 3 07 09 pm" src="https://user-images.githubusercontent.com/361671/39488819-767468f6-4d51-11e8-8808-3ea5b3cd80a3.png">

Clicking this takes you to the place in devtools:

<img width="474" alt="screen shot 2018-05-01 at 3 07 27 pm" src="https://user-images.githubusercontent.com/361671/39488831-7eb86e72-4d51-11e8-9779-44758b239d76.png">

I think this is worth the tradeoffs listed above.
